fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerUniqueIndexTest.java

```diff
@@ -366,4 +371,184 @@ public void resolveUniquenessViolations() {
+        disableAll(indexes);
+        try (OnlineIndexer indexBuilder = OnlineIndexer.newBuilder()
+                .setDatabase(fdb).setMetaData(metaData).setSubspace(subspace)
+                .setTargetIndexes(indexes)
+                .setIndexingPolicy(OnlineIndexer.IndexingPolicy.newBuilder()
+                        .allowUniquePendingState(allow))
+                .build()) {
+            if (allow) {
+                indexBuilder.buildIndex();
+            } else {
+                RecordCoreException e = assertThrows(RecordCoreException.class, indexBuilder::buildIndex);
+                RuntimeException runE = FDBExceptions.wrapException(e);
+                assertNotNull(runE);
+                assertThat(runE, instanceOf(RecordIndexUniquenessViolation.class));
+            }
+        }
+        try (FDBRecordContext context = openContext()) {
+            // unique index with uniqueness violation:
+            if (allow) {
+                assertTrue(recordStore.isIndexReadableUniquePending(indexes.get(0)));
+            } else {
+                assertTrue(recordStore.isIndexWriteOnly(indexes.get(0)));
+            }
+            // non-unique index:
+            assertTrue(recordStore.isIndexReadable(indexes.get(1)));
+            // unique index with on unique number (no duplications):
+            assertTrue(recordStore.isIndexReadable(indexes.get(2)));
+            context.commit();
+        }
+
+        // now try resolving the duplications, and marking readable with another build
+        final Index index = indexes.get(0);
+        try (FDBRecordContext context = openContext()) {
+            Set<Tuple> indexEntries = new HashSet<>(recordStore.scanUniquenessViolations(index)
+                    .map( v -> v.getIndexEntry().getKey() )
+                    .asList().join());
+
+            for (Tuple indexKey : indexEntries) {
+                List<Tuple> primaryKeys = recordStore.scanUniquenessViolations(index, indexKey).map(RecordIndexUniquenessViolation::getPrimaryKey).asList().join();
+                assertEquals(2, primaryKeys.size());
+                recordStore.resolveUniquenessViolation(index, indexKey, primaryKeys.get(0)).join();
+                assertEquals(0, (int)recordStore.scanUniquenessViolations(index, indexKey).getCount().join());
+            }
+
+            for (int i = 0; i < 5; i++) {
+                assertNotNull(recordStore.loadRecord(Tuple.from(i)));
+            }
+            for (int i = 5; i < records.size(); i++) {
+                assertNull(recordStore.loadRecord(Tuple.from(i)));
+            }
+            context.commit();
+        }
+        openSimpleMetaData(hook);
+        try (OnlineIndexer indexBuilder = OnlineIndexer.newBuilder()
+                .setDatabase(fdb).setMetaData(metaData).setIndex(index).setSubspace(subspace)
+                .setIndexingPolicy(OnlineIndexer.IndexingPolicy.newBuilder()
+                        .allowUniquePendingState())
```

Comment:
It would probably be good to test building when an index was in `READABLE_UNIQUE_PENDING`, and still has violations.

Reply:
I'm wondering if there's a way to test if it fails for other reasons, but I can't think of any errors that we could produce, so adding the tests on `FDBRecordStore` that it can change into these states if things are broken might be all we can do, but if you can think of anything that would be great.

Reply:
The only thing I can think about is the config loader crashes. This is heavily used in indexing tests.